### ISSUE TITLE
No implicit conversion of String into Integer error loop when record not in json

### ISSUE THREAD

**Issue by jtblin** [awslabs/aws-fluent-plugin-kinesis#40]
We are getting the following error:

> 2015-10-21 15:14:33 +0000 [warn]: temporarily failed to flush the buffer. next_retry=2015-10-21 15:14:34 +0000 error_class="TypeError"
error="no implicit conversion of String into Integer" plugin_id="object:3fb566013c1c"
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluent-plugin-kinesis-0.3.6/lib/fluent/plugin/out_kinesis.rb:124:in `[]'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluent-plugin-kinesis-0.3.6/lib/fluent/plugin/out_kinesis.rb:124:in `block in write'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluentd-0.12.15/lib/fluent/buffer.rb:117:in `each'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluentd-0.12.15/lib/fluent/buffer.rb:117:in `block in msgpack_each'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluentd-0.12.15/lib/fluent/plugin/buf_file.rb:64:in `open'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluentd-0.12.15/lib/fluent/buffer.rb:114:in `msgpack_each'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluent-plugin-kinesis-0.3.6/lib/fluent/plugin/out_kinesis.rb:123:in `each'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluent-plugin-kinesis-0.3.6/lib/fluent/plugin/out_kinesis.rb:123:in `find_all'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluent-plugin-kinesis-0.3.6/lib/fluent/plugin/out_kinesis.rb:123:in `write'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluentd-0.12.15/lib/fluent/buffer.rb:325:in `write_chunk'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluentd-0.12.15/lib/fluent/buffer.rb:304:in `pop'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluentd-0.12.15/lib/fluent/output.rb:321:in `try_flush'
2015-10-21 15:14:33 +0000 [warn]: /usr/local/bundle/gems/fluentd-0.12.15/lib/fluent/output.rb:140:in `run'

The error is in write on the [following lines](https://github.com/awslabs/aws-fluent-plugin-kinesis/blob/master/lib/fluent/plugin/out_kinesis.rb#L123): `record['data']`

```
data_list = chunk.to_enum(:msgpack_each).find_all{|record|
  unless record_exceeds_max_size?(record['data'])`
```

This is likely due to record not being a json object but an array so `[]` is expecting an integer. The [elasticsearch plugin](https://github.com/uken/fluent-plugin-elasticsearch/blob/master/lib/fluent/plugin/out_elasticsearch.rb#L132) avoids this issue with a guard to check that the record is a hash.

```
next unless record.is_a? Hash
```

On restart of the process, it seems to be looping on this error for error while trying to process the buffer.

**Comment by riywo** [comment]
Hi @jtblin ,

Thank you for the report. Could you share the way to reproduce this?

**Comment by jtblin** [comment]
We think this happened because the buffer got corrupted for some reason when the process crashed, when fluentd restarted it read the corrupted buffer and the plugin encountered the error above when it processed the corrupted buffer, so it is hard to provide a repro case for that as we didn't keep the corrupted buffer when that happened. 

As the code excepts a hash, I think specifically testing for it makes sense to avoid problems with unexpected buffers.

**Comment by riywo** [comment]
Got it. I'll send PR and merge it soon. Thanks!

**Comment by riywo** [comment]
Included in v1.0.0.

**Comment by jtblin** [comment]
Cool, thanks :+1:
